ase/sw/protocol_backend.c

```diff
@@ -219,35 +224,108 @@ void sv2c_script_dex(const char *str)
+		"              write fence before AFU status message. The simulator may\n" \
+		"              be committing AFU transactions out of order.\n"
+
+	#define MEMLINE_ADDR_UNMAPPED_MSG \
+		"@ERROR: ASE has detected a memory %s to an UNMAPPED memory region.\n" \
+		"        Simulation cannot continue, please check the code.\n" \
+		"          Failure @ byte-level phys_addr = 0x%" PRIx64 "\n" \
+		"                    line-level phys_addr = 0x%" PRIx64 "\n" \
+		"                    line-level virtual_addr = %p\n" \
+		"        See ERROR log file ase_memory_error.log and timestamped\n" \
+		"        transactions in ccip_transactions.tsv.\n" \
+		"@ERROR: This most often happens when the application munmaps or frees\n" \
+		"        a region before calling fpgaReleaseBuffer().\n"
+
+	FILE *error_fp = fopen("ase_memory_error.log", "w");
+
+	if (status == HOST_MEM_STATUS_ILLEGAL) {
+		ASE_ERR("\n" MEMLINE_ADDR_ILLEGAL_MSG, access_type, pa, pa >> 6);
+		if (error_fp != NULL) {
+			fprintf(error_fp, MEMLINE_ADDR_ILLEGAL_MSG, access_type, pa, pa >> 6);
+		}
+	} else if (status == HOST_MEM_STATUS_NOT_PINNED) {
+		ASE_ERR("\n" MEMLINE_ADDR_UNPINNED_MSG, access_type, pa, pa >> 6);
+		if (error_fp != NULL) {
+			fprintf(error_fp, MEMLINE_ADDR_UNPINNED_MSG, access_type, pa, pa >> 6);
+		}
+	} else if (status == HOST_MEM_STATUS_NOT_MAPPED) {
+		ASE_ERR("\n" MEMLINE_ADDR_UNMAPPED_MSG, access_type, pa, pa >> 6, va);
+		if (error_fp != NULL) {
+			fprintf(error_fp, MEMLINE_ADDR_UNMAPPED_MSG, access_type, pa, pa >> 6, va);
+		}
+	} else {
+		ASE_ERR("\n @ERROR: Unknown memory reference error.\n");
+	}
+
+	if (error_fp != NULL) {
+		fclose(error_fp);
+	}
+
+	#undef MEMLINE_ADDR_ILLEGAL_MSG
+	#undef MEMLINE_ADDR_UNPINNED_MSG
+	#undef MEMLINE_ADDR_UNMAPPED_MSG
+
+	// Request SIMKILL
+	start_simkill_countdown();
+
+	FUNC_CALL_EXIT;
+}
+
+
+/*
+ * DPI: Write line request (sent to application)
+ */
+void wr_memline_req_dex(cci_pkt *pkt)
 {
 	FUNC_CALL_ENTRY;
 
 	uint64_t phys_addr;
-	uint64_t *wr_target_vaddr = (uint64_t *) NULL;
 	int intr_id;
-	//int ret_fd;
-	/* #ifndef DEFEATURE_ATOMICS */
-	/*   uint64_t *rd_target_vaddr = (uint64_t*)NULL; */
-	/*   long long cmp_qword;  // Data to be compared */
-	/*   long long new_qword;  // Data to be writen if compare passes */
-	/* #endif */
+
+	ase_host_memory_write_req wr_req;
 
 	if (pkt->mode == CCIPKT_WRITE_MODE) {
 		/*
 		 * Normal write operation
 		 * Takes Write request and performs verbatim
 		 */
-		// Get cl_addr, deduce wr_target_vaddr
 		phys_addr = (uint64_t) pkt->cl_addr << 6;
-		wr_target_vaddr =
-			ase_fakeaddr_to_vaddr((uint64_t) phys_addr);
 
 		// Write to memory
-		ase_memcpy(wr_target_vaddr, (char *) pkt->qword,
-			   CL_BYTE_WIDTH);
+		wr_req.req = HOST_MEM_REQ_WRITE_LINE;
+		wr_req.addr = phys_addr;
+		ase_memcpy(wr_req.data, (char *) pkt->qword, CL_BYTE_WIDTH);
+
+		mqueue_send(sim2app_membus_wr_req_tx, (char *) &wr_req, sizeof(wr_req));
```

Comment:
mqueue_send() currently has no return value (void). This is pervasive and should be on the list of things to fix, but is bigger than just this change since it involves figuring out how to tear down the simulation cleanly.